core/src/upgrade.rs

```diff
@@ -145,8 +148,43 @@ pub trait UpgradeInfo {
     fn protocol_info(&self) -> Self::InfoIter;
 }
 
+pub trait UpgradeProtocols {
+    type Iter: Iterator<Item = Protocol>;
```

Comment:
I thought about whether we should expose `multistream_select::Protocol` via `libp2p-core` or whether we should introduce a wrapper type (e.g. also `Protocol`) in `libp2p-core` that wraps `multistream_select::Protocol`.

- Would allow us to move away from multistream-select without a breaking change.
- Would introduce an indirection.

I don't think we will make use of the benefit any time soon. In other words, I don't think we are moving away from multistream-select (e.g. to Protocol Select) any time soon. Thus I favor the way as it is proposed here.

Reply:
Great minds think alike: https://github.com/libp2p/rust-libp2p/issues/3759

Reply:
I've added some more thoughts to #3759. I think for this particular PR, I wouldn't bother. But maybe we can land something on top. I'll add it to the milestone so we get reminded.

Reply:
I've encapsulated the `Protocol` struct now. In a later PR, I want to deal with the error type.